File: mobile/lib/models/calendar.dart
import '../utils/num_utils.dart';

class AcademicCalendar {
  final String id;
  final String? semesterId;
  final String? date;
  final int weekNumber;
  final int dayOfWeek;
  final String? eventType;
  final String? eventPriority;
  final String? eventName;
  final String? eventCode;
  final bool isHoliday;
  final bool isWorkday;
  final bool isTeachingDay;
  final bool isExamDay;
  final bool isAdjusted;
  final String? holidayName;
  final String? holidayType;
  final String? description;
  final String? createdAt;

  AcademicCalendar({
    required this.id,
    this.semesterId,
    this.date,
    this.weekNumber = 0,
    this.dayOfWeek = 0,
    this.eventType,
    this.eventPriority,
    this.eventName,
    this.eventCode,
    this.isHoliday = false,
    this.isWorkday = false,
    this.isTeachingDay = false,
    this.isExamDay = false,
    this.isAdjusted = false,
    this.holidayName,
    this.holidayType,
    this.description,
    this.createdAt,
  });

  factory AcademicCalendar.fromJson(Map<String, dynamic> json) {
    return AcademicCalendar(
      id: json['id']?.toString() ?? '',
      semesterId: json['semesterId']?.toString(),
      date: json['date']?.toString(),
      weekNumber: safeInt(json['weekNumber']),
      dayOfWeek: safeInt(json['dayOfWeek']),
      eventType: json['eventType']?.toString(),
      eventPriority: json['eventPriority']?.toString(),
      eventName: json['eventName']?.toString(),
      eventCode: json['eventCode']?.toString(),
      isHoliday: json['isHoliday'] as bool? ?? false,
      isWorkday: json['isWorkday'] as bool? ?? false,
      isTeachingDay: json['isTeachingDay'] as bool? ?? false,
      isExamDay: json['isExamDay'] as bool? ?? false,
      isAdjusted: json['isAdjusted'] as bool? ?? false,
      holidayName: json['holidayName']?.toString(),
      holidayType: json['holidayType']?.toString(),
      description: json['description']?.toString(),
      createdAt: json['createdAt']?.toString(),
    );
  }
}

class AddHolidayRequest {
  final String date;
  final String name;
  final String? type;
  final bool isWorkday;
  final String? description;

  AddHolidayRequest({
    required this.date,
    required this.name,
    this.type,
    this.isWorkday = false,
    this.description,
  });

  Map<String, dynamic> toJson() => {
        'date': date,
        'name': name,
        'type': type,
        'isWorkday': isWorkday,
        'description': description,
      };
}

class WeekInfo {
  final int currentWeek;
  final String? weekStartDate;
  final String? weekEndDate;
  final bool isTeachingWeek;
  final String? semesterName;

  WeekInfo({
    this.currentWeek = 0,
    this.weekStartDate,
    this.weekEndDate,
    this.isTeachingWeek = false,
    this.semesterName,
  });

  factory WeekInfo.fromJson(Map<String, dynamic> json) {
    return WeekInfo(
      currentWeek: safeInt(json['currentWeek']),
      weekStartDate: json['weekStartDate']?.toString(),
      weekEndDate: json['weekEndDate']?.toString(),
      isTeachingWeek: json['isTeachingWeek'] as bool? ?? false,
      semesterName: json['semesterName']?.toString(),
    );
  }
}

File: uni-app/src/types/calendar.ts
/**
 * 校历相关类型
 * 与后端 AcademicCalendarDto 对齐
 */

export type CalendarEventType =
  | 'Teaching'
  | 'Exam'
  | 'Holiday'
  | 'Registration'
  | 'CourseSelection'
  | 'GradeEntry'
  | 'Sports'
  | 'Activity'
  | 'Maintenance'
  | 'Custom'

export interface AcademicCalendar {
  id: string
  semesterId: string
  date: string
  weekNumber: number
  dayOfWeek: number
  eventType: string
  eventName?: string
  eventPriority?: string
  isHoliday: boolean
  isWorkday: boolean
  isTeachingDay: boolean
  isExamDay?: boolean
  isAdjusted?: boolean
  adjustedFrom?: string
  holidayName?: string
  holidayType?: string
  description?: string
  color?: string
  icon?: string
  affectsCourseSelection: boolean
  affectsScheduling: boolean
  affectsGradeEntry: boolean
  affectsRegistration: boolean
  autoTriggerAction?: string
  triggeredAt?: string
  createdAt: string
  updatedAt?: string
}

export interface UpdateCalendarRequest {
  eventType?: CalendarEventType
  eventName?: string
  isHoliday?: boolean
  isWorkday?: boolean
  isTeachingDay?: boolean
  holidayName?: string
  description?: string
  color?: string
}

export interface HolidayDto {
  id: string
  date: string
  name?: string
  type?: string
  isWorkday: boolean
  description?: string
}

export interface AddHolidayRequest {
  date: string
  name: string
  type?: string
  isWorkday?: boolean
  description?: string
}

export interface AdjustWorkdayRequest {
  date: string
  isWorkday: boolean
  adjustedFrom?: string
  description?: string
}

export interface BusinessPermission {
  isAllowed: boolean
  message: string
  reason: string
  currentDate: string
  weekNumber: number
  dayOfWeek: number
}
